### PR TITLE
Improve Components error handling

### PR DESCRIPTION
**Pull request by pranavkm**
* Change event handlers IHandleEvent, IHandleAfterEvent to be async.
* Return faulted tasks to Renderer instead of handling exceptions in ComponentBase
* Use ILogger in RemoteRenderer, and log to console in WebAssemblyRenderer

Fixes https://github.com/aspnet/AspNetCore/issues/4964

### CODE REVIEW

File: src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs
Comment (pranavkm) [inline src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs]: Unused

File: src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs
Comment (pranavkm) [inline src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs]: We could in theory leave it unhandled in the blazor case. You get a slightly shorter stack trace if you handle it rather than have the runtime catch it.

File: src/Components/Components/src/ComponentBase.cs
Comment (pranavkm) [inline src/Components/Components/src/ComponentBase.cs]: Might be useful to diff this file side-by-side.

File: src/Components/Components/src/Rendering/ComponentState.cs
Comment (pranavkm) [inline src/Components/Components/src/Rendering/ComponentState.cs]: VS tells me to use auto properties instead

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (pranavkm) [inline src/Components/Components/src/Rendering/Renderer.cs]: Using `TaskCancelledException` doesn't always work. For instance, the exception thrown from cancelling a cancellation token results in the context of a Task results in the task transitioning to Canceled, but you still get the `OperationCancelledException`. Inspecting `IsCanceled` seems to be the reliable way to go

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (pranavkm) [inline src/Components/Components/src/Rendering/Renderer.cs]: Mimics the previous behavior where `NotifyRenderCompleted` was an `async void` method

File: src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs
Comment (SteveSandersonMS) [reply src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs]: Well spotted! It used to be the way that .Browser.JS triggered event handlers, but looks like we moved to a different code path when we redid the JS interop.

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/Rendering/Renderer.cs]: Does this method really serve any purpose now? Seems like we could drop it and just make RenderRootComponentAsync be the only one.

File: src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs
Comment (javiercn) [reply src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs]: Lol, I just killed this too as part of my changes

File: src/Components/Components/src/ComponentBase.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/ComponentBase.cs]: Could we rename this to `CallOnParametersSetAsync`? The current name `SetParametersAsyncCore` is super misleading, because it doesn't actually set the parameters - it just notifies the component that the parameters *were* set earlier.

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (javiercn) [reply src/Components/Components/src/Rendering/Renderer.cs]: @SteveSandersonMS I'm fine with it, so would the recommendation for people be to simply not await the returned task if they don't want to wait for the entire render?

File: src/Components/Components/src/ComponentBase.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/ComponentBase.cs]: The simplifications in this file look great.

@javiercn, are you happy that these changes preserve the behavior of the lifecycle async flow you designed for prerendering?

File: src/Components/Components/test/RendererTest.cs
Comment (SteveSandersonMS) [inline src/Components/Components/test/RendererTest.cs]: I'm not certain we really should be awaiting here, or in any of the other tests. That would be a significant change.

Up until now, we've always guaranteed that the event dispatcher and initial re-render is synchronous. It returns a Task now for error-handling purposes, but that shouldn't change the contract that event dispatching and re-render is synchronous.

I think we should explicitly *not* be awaiting so that we can observe the synchronous effects.

Or if you prefer to await in most tests but just add a few tests to show that the dispatch and rerender actually occur synchronously (including for nested components), that would be fine with me too.

File: src/Components/Components/test/RendererTest.cs
Comment (SteveSandersonMS) [inline src/Components/Components/test/RendererTest.cs]: Typo: `HandleAfter` -> `HandleAfterRender`

File: src/Components/Server/src/DependencyInjection/RazorComponentsServiceCollectionExtensions.cs
Comment (SteveSandersonMS) [inline src/Components/Server/src/DependencyInjection/RazorComponentsServiceCollectionExtensions.cs]: Is this safe to do even if the developer is explicitly adding their own logging service elsewhere in Startup.cs? I'm not super familiar with the internals of this.

File: src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj
Comment (SteveSandersonMS) [inline src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj]: I'm curious about what this is for.

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.cs]: @javiercn I guess so. There's no extra advantage in providing an `async void` wrapper around an `async Task` method, as far as I can tell.

File: src/Components/Components/src/ComponentBase.cs
Comment (javiercn) [inline src/Components/Components/src/ComponentBase.cs]: I need to look at this in more detail, I'm not clear if more renders are being introduced.

File: src/Components/Components/test/RendererTest.cs
Comment (SteveSandersonMS) [reply src/Components/Components/test/RendererTest.cs]: In case you're wondering why I think this matters, there are various scenarios in client-side browser code where the behavior is different depending on whether the effects happen synchronously after a user action (such as a click) or if they are async. For example, the "open popup" or "go full screen" APIs are allowed synchronously after a click, but disallowed asynchronously.

Maybe you wouldn't invoke either of those APIs during rendering, but to avoid us having to reason about those fine details on a per-API basis, it's much more reasonable if we guarantee synchrony as much as possible. Asynchrony is hard to reason about and nearly impossible to track down if it starts leaking in.

File: src/Components/Components/src/ComponentBase.cs
Comment (javiercn) [inline src/Components/Components/src/ComponentBase.cs]: This is not correct. We want to trigger a render after calling binding.Invoke has finished executing the sync part and another optional render after the binding has completed executing in case it does actual async work.

This is similar to what we expect happen on SetParametersAsync.
A single render after all sync code has run
An additional render every time we switch between async/sync

File: src/Components/Components/src/Rendering/ComponentState.cs
Comment (javiercn) [reply src/Components/Components/src/Rendering/ComponentState.cs]: @pranavkm Can you not do random cleanups :cry: It distracts from the main point

File: src/Components/Components/src/ComponentBase.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/ComponentBase.cs]: Ah yes!

@pranavkm Would be awesome if we could also have a unit test to represent the flow mentioned by @javiercn here.

File: src/Components/Components/src/ComponentBase.cs
Comment (javiercn) [reply src/Components/Components/src/ComponentBase.cs]: Agree

File: src/Components/Components/src/ComponentBase.cs
Comment (javiercn) [inline src/Components/Components/src/ComponentBase.cs]: Does this handle task.FromException(e)?

File: src/Components/Server/src/DependencyInjection/RazorComponentsServiceCollectionExtensions.cs
Comment (pranavkm) [reply src/Components/Server/src/DependencyInjection/RazorComponentsServiceCollectionExtensions.cs]: It should be, but I don't see MVC or SignalR doing this. I'll move this to the test app.

Comment (javiercn) [reply src/Components/Server/src/DependencyInjection/RazorComponentsServiceCollectionExtensions.cs]: It's likely they don't do because it's already added by something else they need like DataProtection. In any case, it's safe to do this, if it's not idempotent it's a bug in the method implementation. That is the intended behavior. 

You call what you need and it will noop if called more than once.

File: src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj
Comment (pranavkm) [reply src/Components/Server/src/Microsoft.AspNetCore.Components.Server.csproj]: It was for pretty printing the component name. But now that we no longer include it, this reference is gone.

File: src/Components/Components/test/RendererTest.cs
Comment (pranavkm) [reply src/Components/Components/test/RendererTest.cs]: >  I think we should explicitly not be awaiting so that we can observe the synchronous effects.

I think that would require https://github.com/aspnet/AspNetCore/pull/7240/files to go in first

**Comment by pranavkm** [comment]
I addressed changes to the sources, but I'm going to hold off on test changes (which currently don't compile) until we resolve https://github.com/aspnet/AspNetCore/pull/7240/files. There's a @javiercn's PR sets up the tests correctly, and I would need to rebase my changes on top of it to make the correct assertions.

**Comment by SteveSandersonMS** [comment]
@pranavkm OK, sounds good. Please let us know when this is ready for a final review.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs
Comment (javiercn) [inline src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs]: @SteveSandersonMS I think we should simply remove all sync versions of this. As we spoke about in the past, if you don't want to handle errors nor wait for it to finish, don't await the Task

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (javiercn) [inline src/Components/Components/src/Rendering/Renderer.cs]: This try-catch block probably needs to be done before the Task.WhenAll so that we report on the individual exception.

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (javiercn) [inline src/Components/Components/src/Rendering/Renderer.cs]: Can you explain to me the difference between these two? I thought you had to do ExceptionDispatchInfo.Capture to preserve the original callstack

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (javiercn) [inline src/Components/Components/src/Rendering/Renderer.cs]: I think you should do this per individual task, not on the Task.WhenAll as you want to handle exceptions from individual components

File: src/Components/Components/src/Rendering/HtmlRenderer.cs
Comment (javiercn) [inline src/Components/Components/src/Rendering/HtmlRenderer.cs]: I'm all for killing the sync versions of these. @SteveSandersonMS any objection? (We can file a separate issue for this and do it in one go)

File: src/Components/Components/src/ComponentBase.cs
Comment (javiercn) [inline src/Components/Components/src/ComponentBase.cs]: This introduces an extra StateHasChanged event. We want StateHasChanged to be called right before every "true" async operation and after the sum of the sync operations and the "sync" part of an async operation runs.

For example:
OnInit
OnInitAsync (Completely synchronous)
OnParametersSet()
OnParametersSetAsync (Completely synchronous)
StateHasChanged()

OnInit
OnInitAsync (Completely synchronous)
OnParametersSet()
OnParametersSetAsync (Sync)
StateHasChanged()
OnParametersSetAsync (Continuation)
StateHasChanged()

File: src/Components/Components/src/ComponentBase.cs
Comment (javiercn) [inline src/Components/Components/src/ComponentBase.cs]: Can we find a better name for this? At this point is basically awaiting an async task, swallowing cancellations and calling StateHasChanged. 

Could we call this something like CallStateHasChangedOnAsyncCompletion or something like that?

File: src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs
Comment (SteveSandersonMS) [reply src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs]: Sounds good to me.

File: src/Components/Components/src/ComponentBase.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/ComponentBase.cs]: @javiercn Please let us know if you have had a chance to do this.

File: src/Components/Components/src/Rendering/HtmlRenderer.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/HtmlRenderer.cs]: Totally agreed.

File: src/Components/Components/test/RendererTest.cs
Comment (SteveSandersonMS) [inline src/Components/Components/test/RendererTest.cs]: As with the comment above, can we avoid the `await`s in these tests so they are making assertions about synchronous behavior?

**Comment by SteveSandersonMS** [comment]
@pranavkm This is looking great. I appreciate it's ended up being quite complex and fiddly, so thanks for continuing to track down all the details. @javiercn has made some good comments so I'll look forward to seeing the final version of the PR.

**Comment by javiercn** [comment]
@pranavkm Let me know when you need me to look at this again

**Review by pranavkm — COMMENTED** [review]
🆙 📅

File: src/Components/Components/src/ComponentBase.cs
Comment (javiercn) [inline src/Components/Components/src/ComponentBase.cs]: Nit: Is there something specific you can catch in here instead of using task.IsCancelled? It's just weird to read a catch block with a when clause instead of an exception. It's not immediate how the exception thrown and the condition are related.

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/Rendering/Renderer.cs]: Is this new code block meant to be here? I'm guessing we wouldn't really want to throw a completely blank `Exception`. Doesn't the `Debug.Assert` on L157 handle this adequately already?

**Review by SteveSandersonMS — APPROVED** [review]
This is looking solid to me.

@javiercn Could you give sign-off before @pranavkm merges? I know you've had some pretty detailed involvement in the final changes here.

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (pranavkm) [reply src/Components/Components/src/Rendering/Renderer.cs]: Debugging something and forgot to remove it. `Debug.Asserts` just result in xunit crashing the test process and I had a bug here at some point

File: src/Components/Components/src/ComponentBase.cs
Comment (pranavkm) [reply src/Components/Components/src/ComponentBase.cs]: We get a mix of two different exceptions depending on how the Task was cancelled. Throwing a `OperationCanceledException` (`cancelationToken.ThrowIfCancellationRequested()`) does this. The `Task` becomes cancelled, but the exception continues to be `OperationCancelledException`. Alternatively awaiting a cancelled task `Task.FromCanceled` results in a `TaskCancellationException` which also has the same state. Observing the state of the task seems like less work \ easier to grock.

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (pranavkm) [inline src/Components/Components/src/Rendering/Renderer.cs]: @javiercn noticed a blazor test fail and this turned out to be the culprit. `workItem()` returns a `Task` and we weren't awaiting it as part of `InvokeAsync`.

File: src/Components/Components/src/Rendering/Renderer.cs
Comment (pranavkm) [inline src/Components/Components/src/Rendering/Renderer.cs]: Changed this from `async void` to `Task` returning which we ignore. This follows the same pattern we would tell users to do, plus avoids any state machine if everything is synchronous.

**Review by pranavkm — COMMENTED** [review]
🆙 📅

File: src/Components/Components/src/ComponentBase.cs
Comment (pranavkm) [reply src/Components/Components/src/ComponentBase.cs]: See https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-cancellation. The thing is a bit involved.

Comment (javiercn) [reply src/Components/Components/src/ComponentBase.cs]: Can you add a comment? I'm happy if we have a comment so that people can understand it.

Comment (pranavkm) [reply src/Components/Components/src/ComponentBase.cs]: Done

**Comment by pranavkm** [comment]
🆙 📅

**Review by javiercn — APPROVED** [review]
Great improvement!

File: src/Components/Components/src/IHandleAfterRender.cs
Comment (rynowak) [inline src/Components/Components/src/IHandleAfterRender.cs]: This looks like a bad copy-paste

Comment (pranavkm) [reply src/Components/Components/src/IHandleAfterRender.cs]: I couldn't think of anything else to say here so went with the boiler plate. Suggestions?

Comment (rynowak) [reply src/Components/Components/src/IHandleAfterRender.cs]: `asynchronous event handling operation` -> `asynchronous rendering operation`

**Comment by pranavkm** [comment]
Thanks for looking through the PR!

**Comment by chucker** [comment]
Hi,

I saw someone struggle with a dependency injection error in an RC project, where adding an `@inject` statement to their component caused a blank page to render. No error was logged to the browser console or the VS output window.

Will this PR improve this behavior, or should I file a separate issue?

**Comment by pranavkm** [comment]
A new issue with steps to reproduce the error would be great.

**Pull request merged**
